Clean test output for gotestsum

Side effect of #1962:

```
14:42:38  gotestsum: Error: failed to parse test output: rpc error: code = Unknown desc = Error: 
  No such container: 385257478d2fb705cb510d7e30f30640450cdcaa98ba2a021b0816eb4dc6202d: 
  invalid character 'r' looking for beginning of value
14:42:38  Makefile:369: recipe for target 'e2e-generate-xml' failed
```
https://devops-ci.elastic.co/view/cloud-on-k8s/job/cloud-on-k8s-versions-gke/86/execution/node/84/log/

It looks like the test output can be composed of non-test inputs and needs to be cleaned before being passed to gotestsum.

--- a/test/e2e/cmd/run/jsonlog.go
+++ b/test/e2e/cmd/run/jsonlog.go
@@ -1,0 +1,91 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package run
+
+import (
+	"bufio"
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io"
+	"os"
+)
+
+const newLine = byte('\n')
+
+// jsonLog implements a Writer that only emits valid JSON to output.
+type jsonLog struct {
+	out    io.WriteCloser
+	writer *bufio.Writer
+	buf    *bytes.Buffer
+}
+
+func newJSONLogToFile(outFile string) (*jsonLog, error) {
+	f, err := os.Create(outFile)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create JSON log file %s: %w", outFile, err)
+	}
+
+	return newJSONLog(f), nil
+}
+
+func newJSONLog(out io.WriteCloser) *jsonLog {
+	return &jsonLog{
+		out:    out,
+		writer: bufio.NewWriter(out),
+		buf:    new(bytes.Buffer),
+	}
+}
+
+func (jl *jsonLog) Write(p []byte) (int, error) {
+	for i, b := range p {
+		if b == newLine {
+			if err := jl.writeLine(); err != nil {
+				return 0, err
+			}
+			continue
+		}
+		if err := jl.buf.WriteByte(b); err != nil {
+			return i, err
+		}
+	}
+	return len(p), nil
+}
+
+func (jl *jsonLog) writeLine() error {
+	line := jl.buf.Bytes()
+	if json.Valid(line) {
+		if _, err := jl.writer.Write(line); err != nil {
+			return err
+		}
+		if _, err := jl.writer.Write([]byte{newLine}); err != nil {
+			return err
+		}
+	}
+
+	jl.buf.Reset()
+	return nil
+}
+
+func (jl *jsonLog) Close() (err error) {
+	defer func() {
+		e := jl.out.Close()
+		if err == nil {
+			err = e
+		}
+	}()
+
+	// flush internal buffer
+	if err = jl.writeLine(); err != nil {
+		return
+	}
+
+	// flush the writer
+	if err = jl.writer.Flush(); err != nil {
+		return
+	}
+
+	return nil
+}

--- a/test/e2e/cmd/run/jsonlog_test.go
+++ b/test/e2e/cmd/run/jsonlog_test.go
@@ -1,0 +1,45 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package run
+
+import (
+	"bufio"
+	"bytes"
+	"io"
+	"testing"
+
+	"github.com/stretchr/testify/require"
+)
+
+// fakeWriteCloser fakes an io.WriteCloser. (https://github.com/golang/go/issues/22823)
+type fakeWriteCloser struct {
+	io.Writer
+}
+
+func (f *fakeWriteCloser) Close() error {
+	return nil
+}
+
+func TestJSONLog(t *testing.T) {
+	buf := new(bytes.Buffer)
+	jl := newJSONLog(&fakeWriteCloser{Writer: buf})
+
+	w := bufio.NewWriter(jl)
+	_, err := w.WriteString(`
+gibberish
+{"key1": "value1", "key2": "value2"}
+garbage
+{"Time":"2019-10-30T10:52:39.933604025Z","Action":"output","Package":"github.com/elastic/cloud-on-k8s/test/e2e/es","Test":"TestVolumeEmptyDir/Creating_an_Elasticsearch_cluster_should_succeed","Output":"=== RUN   TestVolumeEmptyDir/Creating_an_Elasticsearch_cluster_should_succeed\n"}`)
+
+	require.NoError(t, err)
+	require.NoError(t, w.Flush())
+	require.NoError(t, jl.Close())
+
+	want := `{"key1": "value1", "key2": "value2"}
+{"Time":"2019-10-30T10:52:39.933604025Z","Action":"output","Package":"github.com/elastic/cloud-on-k8s/test/e2e/es","Test":"TestVolumeEmptyDir/Creating_an_Elasticsearch_cluster_should_succeed","Output":"=== RUN   TestVolumeEmptyDir/Creating_an_Elasticsearch_cluster_should_succeed\n"}
+`
+
+	require.Equal(t, want, buf.String())
+}

--- a/test/e2e/cmd/run/run.go
+++ b/test/e2e/cmd/run/run.go
@@ -343,17 +343,18 @@ func (h *helper) streamTestJobOutput(streamStatus chan<- error, client *kubernet
 	}
 	defer stream.Close()
 
-	writer := io.MultiWriter(os.Stdout)
+	outputs := []io.Writer{os.Stdout}
 	if h.logToFile {
-		f, err := os.Create(testsLogFile)
+		jl, err := newJSONLogToFile(testsLogFile)
 		if err != nil {
-			log.Error(err, "Can't create file for test output")
+			log.Error(err, "Failed to create log file for test output")
 			return
 		}
-		defer f.Close()
-		writer = io.MultiWriter(writer, f)
+		defer jl.Close()
+		outputs = append(outputs, jl)
 	}
 
+	writer := io.MultiWriter(outputs...)
 	var buffer [logBufferSize]byte
 	if _, err := io.CopyBuffer(writer, stream, buffer[:]); err != nil {
 		if err == io.EOF {

